### PR TITLE
feat: handle dependency gate cycle recovery and ordering

### LINKED ISSUE #356: Dependency gate slice 3/3: cycle handling, not_planned re-Triage, tie-breaking + CONTEXT.md (oracle review required)
## Parent

#351

## What to build

Tracer-bullet slice 3 of 3 for the dependency-aware Dispatch PRD. After slice #355, Coordinator's Dispatch is gated by `blocked_by` on the happy path. This slice adds the **failure-and-edge cases**: cycle detection action paths, re-Triage on `not_planned`/`duplicate` blocker close, and sub-issue list ordering when multiple Issues unblock simultaneously. CONTEXT.md is updated with the new vocabulary.

This slice is **authority-bearing**. **Mandatory `@oracle` review per AGENTS.md applies — the cycle-detected and `not_planned`-closed-blocker paths produce label removals on multiple Issues, and the tie-breaking rule decides Dispatch ordering.** The implementation PR description must explicitly request `@oracle` review and cite ADR-0002 and ADR-0004.

Concretely:

- **Cycle handling in `internal/coordinator/runner`.** When `depgraph.Cycles()` returns non-empty:
  - For every Issue in any cycle: remove `triaged` and `dispatch/*` labels (returns the Issue to the re-Triage candidate set).
  - Post one stamped comment per Issue with a new `<!-- looper:coordinator:cycle -->` self-dedup marker, listing the cycle path (e.g. "This Issue is part of a `blocked_by` cycle: #A → #B → #A. Coordinator has returned the cycle members to the re-Triage candidate set. Resolve the cycle by editing the `blocked_by` relationships, then re-Triage will form a fresh Disposition.").
  - The marker convention follows the existing `<!-- looper:coordinator:* -->` pattern from CONTEXT.md and PRD #334. Marker presence makes re-runs idempotent — re-running on the same cycle does not double-post.

- **`not_planned` / `duplicate` blocker close handling in `internal/coordinator/runner`.** When `depgraph` reports a blocker as `requires-re-triage=true` (i.e. closed with `state_reason==not_planned` or `state_reason==duplicate`):
  - For every direct dependent of that blocker: remove `triaged` and `dispatch/*` labels.
  - **No comment is posted** — the close itself is the human signal; re-Triage will produce a fresh Triage comment naturally.
  - Idempotent on retry: if `triaged` and `dispatch/*` are already absent, do nothing.

- **Tie-breaking on simultaneous unblock in `internal/coordinator/runner`.** When multiple sibling Issues become Dispatchable in the same tick:
  1. Build the candidate set of newly-Dispatchable Issues.
  2. Dispatch up to the existing concurrency budget, in **sub-issue list order under the parent**. The order is read via the `sub_issues` REST wrapper (added in slice #354) of each candidate's parent Issue.
  3. Issues without a parent sub-issue listing fall back to ascending Issue number.
  4. The remainder is queued for next tick — no private queue is materialized; next tick will re-derive the same order from the same GitHub state. Preserves ADR-0001's stateless property.

- **Reopened-blocker policy** (locked from PRD #351, implemented as runner behavior, not new module logic):
  - If a blocker is reopened **after** its dependent already received the Trigger label (Worker possibly running): do nothing. Reopen affects only future Dispatch decisions.
  - If the dependent does not yet have a Trigger label: next tick the dependency gate will hold again, blocking re-Dispatch — this falls out naturally from the gate logic in slice #355 with no special-case code.
  - This is documented in the slice and verified by an integration test, but adds no runtime branching.

- **Unit tests for `runner` extension** (decision-matrix driven where pure, integration-style otherwise):
  - Cycle detection action: graph with two-node cycle A↔B → both A and B get `triaged` + `dispatch/*` removed; both get one cycle comment with `<!-- looper:coordinator:cycle -->` marker
  - Cycle detection action: graph with three-node cycle A→B→C→A → all three Issues get the same treatment
  - Cycle detection idempotency: running on the same cycle a second tick (marker comment already present) does not double-post
  - `not_planned` blocker action: blocker closed as `not_planned`, dependent has `triaged` + `dispatch/*` → both labels removed from dependent; no comment posted
  - `duplicate` blocker action: same as `not_planned`
  - `not_planned` action idempotency: dependent labels already absent → no-op
  - Tie-breaking: three sub-issues all become Dispatchable, concurrency cap = 2 → first two in sub-issue list order Dispatched; third remains for next tick
  - Tie-breaking fallback: candidates without a parent sub-issue listing → ascending Issue-number order
  - Reopened-blocker: Issue B has Trigger label and blocker A reopens → no action taken on B (in-flight respected)
  - Reopened-blocker: Issue B does not yet have Trigger label and blocker A reopens → next tick B is correctly held by the dependency gate (no special-case code; falls out of slice #355's gate)

- **Integration tests via `fake-gh`** — three scenarios:
  1. Two-node cycle (A blocked-by B, B blocked-by A). Tick: both A and B return to re-Triage candidate set; each receives one cycle comment.
  2. Blocker closed as `not_planned`. Tick: dependent's `triaged` and `dispatch/*` removed; no cycle comment posted.
  3. Three sub-issues unblock simultaneously, concurrency cap = 2. Tick: first two in sub-issue list order Dispatched; third remains for next tick.

- **CONTEXT.md updates.** Extend the existing glossary with three new entries, paralleling the structure used for Triage / Disposition / Dispatch / Trigger label / Veto signal:
  - **Blocker**: An Issue listed in another Issue's GitHub-native `blocked_by` set. The blocker's `state` and `state_reason` are the named Authority (with `blocked_by` itself) for the dependency gate.
  - **Dependency gate**: The Dispatch precondition that all blockers be `state==closed AND state_reason==completed`. Held when any blocker is open or closed-not-completed; released when all blockers satisfy the condition.
  - **Ready set**: The subset of tracked Issues whose dependency gate currently holds — i.e. Issues that may be Dispatched this tick subject to existing PRD #334 conditions.

- **Update CONTEXT.md "Relationships" and "Example dialogue" sections** to mention the dependency gate where natural — at minimum, add one bullet to Relationships ("A **Coordinator** consults the **Dependency gate** before performing **Dispatch** when `roles.coordinator.dependencies.enabled = true`") and one Q/A turn to the Example dialogue covering the dependency gate.

This slice does NOT introduce new database tables, persisted state, or webhook subscriptions. The cycle action's marker comment is durable, public, and idempotent. The `not_planned` action is a label-only change (durable, public, idempotent). Tie-breaking is re-derived from GitHub state every tick. ADR-0001 is preserved.

Use domain language from `CONTEXT.md` consistently — including the new vocabulary added in this slice.

## Acceptance criteria

- [ ] Cycle detection: every Issue in any cycle returned by `depgraph.Cycles()` has `triaged` and `dispatch/*` removed and receives one stamped comment with the new `<!-- looper:coordinator:cycle -->` marker listing the cycle path
- [ ] Cycle comments are idempotent on retry (marker presence prevents double-post)
- [ ] `not_planned` blocker close: every direct dependent has `triaged` and `dispatch/*` removed; no comment posted (the close is the human signal)
- [ ] `duplicate` blocker close: same treatment as `not_planned`
- [ ] `not_planned` / `duplicate` actions are idempotent (dependent labels already absent → no-op)
- [ ] Tie-breaking on simultaneous unblock dispatches up to the concurrency cap in sub-issue list order under the parent
- [ ] Tie-breaking fallback (no parent sub-issue listing) uses ascending Issue-number order
- [ ] Tie-breaking overflow is naturally re-derived next tick — no private queue persisted
- [ ] Reopened blocker after dependent received Trigger label: no action taken on the in-flight dependent
- [ ] Reopened blocker before dependent received Trigger label: dependency gate from slice #355 holds correctly with no new special-case code
- [ ] Unit-test matrix above is fully covered for the runner extensions
- [ ] Three `fake-gh` integration tests cover: cycle detection, `not_planned` re-Triage, simultaneous-unblock tie-breaking
- [ ] CONTEXT.md is extended with **Blocker**, **Dependency gate**, **Ready set** entries, plus updates to Relationships and Example dialogue sections
- [ ] PR description **explicitly requests `@oracle` review** per AGENTS.md, citing ADR-0002 and ADR-0004
- [ ] PR description links PRD #351, slice #354, slice #355
- [ ] No new database tables, no schema migrations, no new persisted fields — Coordinator remains stateless per ADR-0001
- [ ] `go test ./...` passes; `go vet ./...` clean; `gofmt -l .` empty

## Blocked by

#355

### PR DESCRIPTION
**Pull request by mrcfps**
## Summary
- add coordinator dependency-gate edge-case handling for `blocked_by` cycles, `not_planned`/`duplicate` blocker closes, and simultaneous-unblock tie-breaking in sub-issue order with issue-number fallback
- update coordinator unit/fake-gh coverage plus CONTEXT.md vocabulary so dispatch gating, ready-set semantics, and cycle comment behavior stay stateless and GitHub-derived

## Why
- The authority for these actions is GitHub-native `blocked_by` plus blocker `state`/`state_reason`; when that authority shows a cycle or a blocker closed without completion, Dispatch cannot safely proceed on stale coordinator output.
- A simpler Dispatch-only hold is insufficient here because cycle members and dependents of `not_planned`/`duplicate` blockers need a fresh Disposition, not just a paused Dispatch, while tie-breaking must be re-derived from GitHub sub-issue state instead of a private queue.
- Cost: extra per-tick GitHub reads, public cycle comments, and label removals on affected Issues; no new database tables, persisted fields, or private ordering state are introduced, preserving ADR-0001.

## Validation
- `gofmt -l .`
- `go vet ./...`
- `go test ./...`

## References
- PRD: #351
- Slice 1: #354
- Slice 2: #355
- ADR-0002: `docs/adr/0002-coordinator-authority-via-durable-labels.md`
- ADR-0004: `docs/adr/0004-dependency-gate-via-github-native-blocked-by.md`

## Review request
- @oracle review required per AGENTS.md for this authority-bearing Coordinator change.

Closes #356

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=worker · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

### CODE REVIEW

File: internal/coordinator/runner.go [hunk 1]
```diff
@@ -465,6 +440,313 @@ func (r *Runner) applyDispatchAction(ctx context.Context, repo string, cwd strin
...
+		if !dependencyTrackedIssue(item.issue, triageCfg.TriagedLabel, dispatchCfg) {
+			continue
+		}
+		ref := depgraph.IssueRef{Repo: normalizeDependencyRepo(repo), Number: item.issue.Number}
+		trackedRefs = append(trackedRefs, ref)
+		state.tracked[ref] = item
+		state.trackedIssueByNumber[item.issue.Number] = ref
+		snapshot.Issues[ref] = depgraph.IssueState{State: item.detail.State, StateReason: item.detail.StateReason}
+		blockers, err := r.github.ListBlockedByIssues(ctx, githubinfra.ViewIssueInput{Repo: repo, IssueNumber: item.issue.Number, CWD: cwd})
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (mrcfps) [reply internal/coordinator/runner.go]: @chatgpt-codex-connector This has been fixed in 8386e02.

I routed buildDependencyState through a retrying ListBlockedByIssues helper in internal/coordinator/runner.go and added a regression test in internal/coordinator/runner_actions_test.go to confirm transient blocked_by read failures are retried.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by mrcfps** [comment]
**Looper fixer round complete** — 8386e02

- ✅ Review comment on `internal/coordinator/runner.go` (@chatgpt-codex-connector) — [thread](https://github.com/nexu-io/looper/pull/380#discussion_r3253185144)
  - I routed buildDependencyState through a retrying ListBlockedByIssues helper in internal/coordinator/runner.go and added a regression test in internal/coordinator/runner_actions_test.go to confirm transient blocked_by read failures are retried.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>
